Implement `MultiBatchVmapTransform::logicalToPhysical(TensorList)`

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #41943 vmap fallback kernel
* **#41942 Implement `MultiBatchVmapTransform::logicalToPhysical(TensorList)`**
* #41941 Optimize alignBatchDimsAtFront
* #41940 Fix const correctness for VmapPhysicalView struct methods

This function:
- permutes all batch dims to the front of the tensors
- aligns all the batch dims to the collective levels of all the tensors
- expands all of the batch dims such that they are present in each of
the result tensors

This function is useful for the next diff up on the stack (which is
implementing a fallback kernel for BatchedTensor). It's also useful in
general for implementing batching rules on operators that take in
multiple batch dimensions at the front of each tensor (but we don't have
too many of those in PyTorch).

Test Plan:
- `./build/bin/vmap_test`

Differential Revision: [D22764104](https://our.internmc.facebook.com/intern/diff/D22764104)

zou3519 committed:
Implement `MultiBatchVmapTransform::logicalToPhysical(TensorList)`

This function:
- permutes all batch dims to the front of the tensors
- aligns all the batch dims to the collective levels of all the tensors
- expands all of the batch dims such that they are present in each of
the result tensors

This function is useful for the next diff up on the stack (which is
implementing a fallback kernel for BatchedTensor). It's also useful in
general for implementing batching rules on operators that take in
multiple batch dimensions at the front of each tensor (but we don't have
too many of those in PyTorch).

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

zou3519 committed:
Update on "Implement `MultiBatchVmapTransform::logicalToPhysical(TensorList)`"


This function:
- permutes all batch dims to the front of the tensors
- aligns all the batch dims to the collective levels of all the tensors
- expands all of the batch dims such that they are present in each of
the result tensors

This function is useful for the next diff up on the stack (which is
implementing a fallback kernel for BatchedTensor). It's also useful in
general for implementing batching rules on operators that take in
multiple batch dimensions at the front of each tensor (but we don't have
too many of those in PyTorch).

Test Plan:
- `./build/bin/vmap_test`

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/test/vmap_test.cpp:
```diff
@@ -850,4 +850,217 @@ TEST(VmapTest, TestBatchedTensorPermute) {
   }
 }
 
+static void checkMultiBatchVmapTransform(TensorList inputs, TensorList expected_outputs) {
```
NB: Most of this PR is tests! There are only 60 non-test line changes

ezyang commented on aten/src/ATen/test/vmap_test.cpp:
```diff
@@ -850,4 +850,217 @@ TEST(VmapTest, TestBatchedTensorPermute) {
+    int64_t B0 = 5, B1 = 7;
+    Tensor x = at::randn({2, B0, 3, B1});
+    Tensor y = at::randn({B1, 2, 3, B0});
+    Tensor batched_x = makeBatched(x, {{0, 1}, {1, 3}});
```
Not the fault of this PR, but I get a pretty bad case of argument blindness with `{{0, 1}, {1, 3}}` here (takes me a moment to translate this into "batch dim 0 at position 1; batch dim 1 at position 3".

ezyang replied:
I see later expands that look like this lol

zou3519 replied:
Sorry that is just me being lazy. I get argument blindness over that too. I've changed all instances of e.g. `{{0, 1}, {1, 3}}` to have comments that show what the dims are:
```
{{/*lvl*/0, /*dim*/3}, {/*lvl*/1, /*dim*/3}}
```

ezyang commented on aten/src/ATen/test/vmap_test.cpp:
```diff
@@ -850,4 +850,217 @@ TEST(VmapTest, TestBatchedTensorPermute) {
+    // Check that batch dims get moved to the front
+    int64_t B0 = 5, B1 = 7;
+    Tensor x = at::randn({2, B0, 3, B1});
+    Tensor y = at::randn({B1, 2, 3, B0});
```
The B0 and B1 convention here makes the intention super clear, thank you!

ezyang commented on aten/src/ATen/test/vmap_test.cpp:
```diff
@@ -850,4 +850,217 @@ TEST(VmapTest, TestBatchedTensorPermute) {
+
+    checkMultiBatchVmapTransform(
+        {batched_x, batched_y},
+        {x.permute({1, 3, 0, 2}), y.permute({3, 0, 1, 2})});
```
NGL, probably would have been easier if we just wrote `B0, B1, 2, 3` as the expected result (and computed the permutation implicitly)

zou3519 replied:
This is where that `torch.moveaxis` PR comes in handy. I've changed the instances of permute introduced in this PR to use `at::movedim`, which is easier to read IMO than permute.

ezyang reviewed (approved):
Maybe a plain English description of the logicalToPhysical algorithm in code would be helpful, but it all looks basically good!

zou3519 committed:
Update on "Implement `MultiBatchVmapTransform::logicalToPhysical(TensorList)`"


This function:
- permutes all batch dims to the front of the tensors
- aligns all the batch dims to the collective levels of all the tensors
- expands all of the batch dims such that they are present in each of
the result tensors

This function is useful for the next diff up on the stack (which is
implementing a fallback kernel for BatchedTensor). It's also useful in
general for implementing batching rules on operators that take in
multiple batch dimensions at the front of each tensor (but we don't have
too many of those in PyTorch).

Test Plan:
- `./build/bin/vmap_test`

Differential Revision: [D22764104](https://our.internmc.facebook.com/intern/diff/D22764104)

[ghstack-poisoned]
